### DISCUSSION TITLE
use unsupervised GraphSAGE(SAGEConv) to classify nodes when only edges have features?

### DISCUSSION THREAD

**Discussion by kokifish** [pyg-team/pytorch_geometric#3544] [Q&A] [answered]
Is there any way to use unsupervised GraphSAGE to classify nodes when only edges have features?
In more detail: 
- nodes: without features
- edges: directed with features
- goal: node classification
When node A sampling A's neighbour node B, B's features are the features of directed edge A->B.
When node C sampling C's neighbour node B, B's features are the features of directed edge C->B.
Node A/C can sample node B only if directed edge A/C->B exits. 

In other words, I have a Data like:
Data(x=[num_nodes, target_embed_dim], edge_index=[2, num_edges], edge_attr=[num_edges, num_edge_features])
where x is a tensor of all zeros.
Can I do this by reimplementing SAGEConv? If yes, how to reimplement it?

**Comment by rusty1s** [comment, marked as answer]
There are two options to tackle this:
1. You need to utilize a GNN tack can incorporate edge features, e.g., `TransformerConv` (see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html)).
2. You need to first convert your edge features into node-level ones, e.g. via `scatter_mean(edge_attr, edge_index[1], ...)`. This can potentially also be learned as part of a GNN pipeline (`scatter_mean(self.mlp(edge_attr), ...)`)

**Reply by kokifish** [reply]
Thank you so much for your help!
However, for my task, nodes are similar to having multiple roles. For example, for central node A, when B is sampled by central node A, the feature of node B is the feature on the directed edge A->B, instead of using The features from other nodes to node B, that is, the feature of directed edge notA->B will not be used.
If I understand correctly, in the above case, when the central node A samples the neighbor node B, the 1. and 2.points will use the features on the edge notA->B.
Do I understand it right? Are there other ways to meet my needs? Or reimplement `SAGEConv`?

**Reply by rusty1s** [reply]
The `NeighborLoader` derived from `GraphSAGE` already performs a filtering of edge features, i.e it only holds the edge features for neighbors that are present in the sampled graph. As such, I believe what you have in mind will still work.

**Reply by kokifish** [reply]
Thank you for your patience.

If I understand correctly. Filtering of edge features means [`filter_data`](https://github.com/pyg-team/pytorch_geometric/blob/1d0efa849c89501799d0965ac83f0c8d7a1d6dc2/torch_geometric/loader/neighbor_loader.py#L202), more precisely, [`filter_edge_store_`](https://github.com/pyg-team/pytorch_geometric/blob/1d0efa849c89501799d0965ac83f0c8d7a1d6dc2/torch_geometric/loader/utils.py#L86). Thus the `Data` return by `NeighborLoader` only holds the edge features for neighbours that are present in the sampled graph(`Data`).

My doubt is whether `SAGEConv`'s `forward`([`propagate() in forward`](https://github.com/pyg-team/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/torch_geometric/nn/conv/sage_conv.py#L69)) can only aggregate the node features([`Data.x` as used here](https://github.com/pyg-team/pytorch_geometric/blob/4bd1821b72039cb6beb96041bc9f18b2c4af36bb/examples/graph_sage_unsup.py#L86)) of neighbouring nodes, and cannot use edge features(`Data.edge_attr`) as node features.

Do I have to reimplement the forward [`propagate()`](https://github.com/pyg-team/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/torch_geometric/nn/conv/sage_conv.py#L69) for passing `edge_attr` and [`message_and_aggregate()`](https://github.com/pyg-team/pytorch_geometric/blob/e6b8d6427ad930c6117298006d7eebea0a37ceac/torch_geometric/nn/conv/sage_conv.py#L84) for using `edge_attr` for `matmul` in SAGEConv?
Very much looking forward to your reply.

**Reply by rusty1s** [reply]
Yes, that is correct. `SAGEConv` cannot handle edge features by definition. As an alternative, you can use any `GNN` layer that can handle edge features, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html). If you want to incorporate edge features into `SAGEConv`, this will look as follows (note that `message_and_aggregate` cannot be implemented in this scenario):
```python
def forward(self, x, edge_index, edge_attr):
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
    return self.lin2(x) + out

def message(self, x_j, edge_attr):
    return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
```

**Reply by kokifish** [reply]
Thank you very much for your patience.
It works for me. Thanks again!

**Reply by downeykking** [reply]
> Yes, that is correct. `SAGEConv` cannot handle edge features by definition. As an alternative, you can use any `GNN` layer that can handle edge features, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html). If you want to incorporate edge features into `SAGEConv`, this will look as follows (note that `message_and_aggregate` cannot be implemented in this scenario):
> 
> ```python
> def forward(self, x, edge_index, edge_attr):
>     out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
>     return self.lin2(x) + out
> 
> def message(self, x_j, edge_attr):
>     return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
> ```

Hello, I am confused by this
```python
def forward(self, x, edge_index, edge_attr):
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
    return self.lin2(x) + out

def message(self, x_j, edge_attr):
    return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
```
when we add edge_attr in sageconv, should this be changed to 
```python 
def forward(self, x, edge_index, edge_attr):
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
    return self.lin2(x) + out

def message(self, x_j, edge_attr_j):
    return self.lin1(torch.cat([x_j, edge_attr_j], dim=-1))
```
because the `x_j` has been chosen by `x[row]`, so maybe `edge_attr` should be transfered to `edge_attr_j`?

**Reply by downeykking** [reply]
> > Yes, that is correct. `SAGEConv` cannot handle edge features by definition. As an alternative, you can use any `GNN` layer that can handle edge features, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html). If you want to incorporate edge features into `SAGEConv`, this will look as follows (note that `message_and_aggregate` cannot be implemented in this scenario):
> > ```python
> > def forward(self, x, edge_index, edge_attr):
> >     out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
> >     return self.lin2(x) + out
> > 
> > def message(self, x_j, edge_attr):
> >     return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
> > ```
> 
> Hello, I am confused by this
> 
> ```python
> def forward(self, x, edge_index, edge_attr):
>     out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
>     return self.lin2(x) + out
> 
> def message(self, x_j, edge_attr):
>     return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
> ```
> 
> when we add edge_attr in sageconv, should this be changed to
> 
> ```python
> def forward(self, x, edge_index, edge_attr):
>     out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
>     return self.lin2(x) + out
> 
> def message(self, x_j, edge_attr_j):
>     return self.lin1(torch.cat([x_j, edge_attr_j], dim=-1))
> ```
> 
> because the `x_j` has been chosen by `x[row]`, so maybe `edge_attr` should be transfered to `edge_attr_j`?

I got it. This is right.
```python
def forward(self, x, edge_index, edge_attr):
    out = self.propagate(edge_index, x=x, edge_attr=edge_attr):
    return self.lin2(x) + out

def message(self, x_j, edge_attr):
    return self.lin1(torch.cat([x_j, edge_attr], dim=-1))
```
Thanks!
